### PR TITLE
ARROW-3770: [C++] Validate schema for each table written with parquet::arrow::FileWriter

### PR DESCRIPTION
**Pull request by bkietz**
`parquet::arrow::FileWriter` now stores the schema it is created with and checks it against tables passed to `WriteTable`.

A translation of pyarrow/tests/test_parquet.py::test_validate_schema_write_table into C++ is also added to parquet-arrow-reader-writer-test to verify mismatched schemas emit an error.

### CODE REVIEW

File: cpp/src/parquet/arrow/arrow-reader-writer-test.cc [hunk 1]
```diff
@@ -140,6 +140,69 @@ LogicalType::type get_logical_type(const ::DataType& type) {
   return LogicalType::NONE;
 }
 
+std::shared_ptr<::arrow::DataType> get_arrow_type(const PrimitiveNode& p) {
```
Comment (wesm) [inline cpp/src/parquet/arrow/arrow-reader-writer-test.cc]: Isn't this https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/schema.cc#L168?

File: cpp/src/parquet/arrow/arrow-reader-writer-test.cc [hunk 2]
```diff
@@ -441,6 +504,16 @@ static std::shared_ptr<GroupNode> MakeSimpleSchema(const ::DataType& type,
   return std::static_pointer_cast<GroupNode>(node_);
 }
 
+std::shared_ptr<::arrow::Schema> ToArrowSchema(const GroupNode& schema_node) {
```
Comment (wesm) [inline cpp/src/parquet/arrow/arrow-reader-writer-test.cc]: Is this https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/schema.cc#L336?

Comment (bkietz) [reply cpp/src/parquet/arrow/arrow-reader-writer-test.cc]: Thanks, that's handy. I'll refactor to use that

**Review by wesm — APPROVED** [review]
+1. Thanks @bkietz!

**Pull request closed without merging**